### PR TITLE
New gas model, storage gas price update and conservation checks

### PR DESCRIPTION
**Pull request by dariorussi**
## Description 

This introduces the new gas model and conservation checks. It also fixes the storage cost to be 76.
The new and old gas model live side by side and are switched over a protocol config gas version. The code tries to fork the code for old and new as close the top of the call stack as possible, to make sure there is as close to 0 interaction as possible between old and new code.
We also have new value for max and min budget all expressed in MIST.
Bucket tables have been changed too.

## Test Plan 

Existing tests

---
If your changes are not user-facing and not a breaking change, you can skip the following section. Otherwise, please indicate what changed, and then add to the Release Notes section as highlighted during the release process.

### Type of Change (Check all that apply)

- [ ] user-visible impact
- [ ] breaking change for a client SDKs
- [ ] breaking change for FNs (FN binary must upgrade)
- [ ] breaking change for validators or node operators (must upgrade binaries)
- [ ] breaking change for on-chain data layout
- [ ] necessitate either a data wipe or data migration

### CODE REVIEW

File: crates/sui-protocol-config/src/snapshots/sui_protocol_config__test__version_1.snap [hunk 1]
```diff
@@ -6,6 +6,7 @@ version: 1
 feature_flags:
   package_upgrades: false
   commit_root_state_digest: false
+  conservation_checks: false
```
Comment (lxfind) [inline crates/sui-protocol-config/src/snapshots/sui_protocol_config__test__version_1.snap:9]: This shouldn't belong here?

Comment (dariorussi) [reply crates/sui-protocol-config/src/snapshots/sui_protocol_config__test__version_1.snap:9]: yes it does not and should remove it

File: crates/sui-protocol-config/src/snapshots/sui_protocol_config__test__version_2.snap [hunk 2]
```diff
@@ -1,13 +1,13 @@
 ---
 source: crates/sui-protocol-config/src/lib.rs
+assertion_line: 1640
```
Comment (lxfind) [inline crates/sui-protocol-config/src/snapshots/sui_protocol_config__test__version_2.snap:3]: What is this?

File: crates/sui-types/src/gas_model/gas_v2.rs [hunk 3]
```diff
@@ -0,0 +1,501 @@
...
+                ExecutionErrorKind::InsufficientGas.into()
+            })
+    }
+
+    fn charge_storage_mutation(
+        &mut self,
+        new_size: usize,
+        storage_rebate: u64,
+    ) -> Result<u64, ExecutionError> {
+        // TODO: this is exclusively called from testing as it breaks too much
```
Comment (lxfind) [inline crates/sui-types/src/gas_model/gas_v2.rs:364]: What does this TODO mean?

File: crates/sui-types/src/gas_model/gas_v2.rs [hunk 4]
```diff
@@ -0,0 +1,501 @@
...
+        // if we fork tests between old and new, revisit
+        let size = self.track_storage_mutation(new_size, storage_rebate);
+        self.charge_storage_and_rebate()?;
+        Ok(size)
+        // Err(ExecutionError::invariant_violation(
```
Comment (lxfind) [inline crates/sui-types/src/gas_model/gas_v2.rs:369]: Remove?

File: crates/sui-types/src/temporary_store.rs [hunk 5]
```diff
@@ -901,6 +896,215 @@ impl<S: ObjectStore> TemporaryStore<S> {
...
+                }
+            }
+
+            // On error we need to dump writes, deletes, etc before charging storage gas
+            if execution_result.is_err() {
+                self.reset(gas, gas_status);
+            }
+        }
+
+        // collect and charge storage cost
```
Comment (lxfind) [inline crates/sui-types/src/temporary_store.rs:943]: The comment doesn't seem to describe what happens inside

**Review by lxfind — COMMENTED** [review]
Not sure how to review the detailed changes since it's not clear what was changed in each function compared to the old version.

**Review by lxfind — DISMISSED** [review]
`GasCostSummary::new` still set `non_refundable_storage_fee` to 0, which seems wrong to me.
Could we also add a test to make sure that we track non_refundable_storage_fee properly?

File: crates/sui-types/src/gas_model/gas_v2.rs (hunk 3 above)
Comment (dariorussi) [reply crates/sui-types/src/gas_model/gas_v2.rs:364]: a bunch of tests are calling that API so for now I am leaving that around even though it is never called at runtime.
I'll fix it once we are all off this collective effort to fix test

File: crates/sui-types/src/gas_model/gas_v2.rs (hunk 4 above)
Comment (dariorussi) [reply crates/sui-types/src/gas_model/gas_v2.rs:369]: will re-enable once I fix the tests call

**Comment by dariorussi** [comment]
> > `GasCostSummary::new` still set `non_refundable_storage_fee` to 0, which seems wrong to me. Could we also add a test to make sure that we track non_refundable_storage_fee properly?
> 
> I think that is a GasCostSummary that is not called, the one in gas_v2.rs is the one that matters, but let me look and remove code that is not invoked as I believe there is still some

it is actually called from some code and a bunch of tests and I'll fix it.

File: crates/sui-types/src/gas_model/gas_v2.rs [hunk 6]
```diff
@@ -0,0 +1,501 @@
...
+    /// storage_rebate is the total storage rebate (in Sui) accumulated in this transaction.
+    /// This is an accumulator computed at the end of execution while determining storage charges.
+    /// It is the sum of all `storage_rebate` of all objects mutated or deleted during
+    /// execution. The value is in Sui.
+    storage_rebate: u64,
+    // storage rebate rate as defined in the ProtocolConfig
+    rebate_rate: u64,
+    /// Amount of storage rebate accumulated when we are running in unmetered mode (i.e. system transaction).
+    /// This allows us to track how much storage rebate we need to retain in system transactions.
+    unmetered_storage_rebate: u64,
```
Comment (lxfind) [inline crates/sui-types/src/gas_model/gas_v2.rs:181]: This value seems never updated in this PR. Missed in the merge?

Comment (dariorussi) [reply crates/sui-types/src/gas_model/gas_v2.rs:181]: possibly

Comment (dariorussi) [reply crates/sui-types/src/gas_model/gas_v2.rs:181]: that was it! thanks!

File: crates/sui-core/src/checkpoints/mod.rs [hunk 7]
```diff
@@ -835,6 +835,8 @@ impl CheckpointBuilder {
                 previous_gas_costs.computation_cost + current_gas_costs.computation_cost,
                 previous_gas_costs.storage_cost + current_gas_costs.storage_cost,
                 previous_gas_costs.storage_rebate + current_gas_costs.storage_rebate,
+                previous_gas_costs.non_refundable_storage_fee
```
Comment (andll) [inline crates/sui-core/src/checkpoints/mod.rs:838]: It looks like we could just make an `Add` implementation to the GasCostSummary and change this code to `previous_gas_costs + current_gas_costs`

Comment (dariorussi) [reply crates/sui-core/src/checkpoints/mod.rs:838]: I'll consider thanks @andll 
for this PR (that is going crazy) maybe I'll stop making changes for now

**Review by sblackshear — APPROVED** [review]

**Pull request merged**